ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -0,0 +1,3018 @@
+            keyboard[Key.font] = font
+          }
+          keyboardsInfo![kbID!] = keyboard
+        }
+
+        let dictKey = "\(langId!)_\(kbID!)"
+        if keyboardsDictionary[dictKey] == nil {
+          let oskFont = kbDict[Key.oskFont] as? [AnyHashable: Any]
+          var dict = [
+            Key.keyboardId: (kbDict[Key.id] as? String)!,
+            Key.languageId: (langDict[Key.id] as? String)!,
+            Key.keyboardName: (kbDict[Key.name] as? String)!,
+            Key.languageName: (langDict[Key.name] as? String)!,
+            Key.keyboardVersion: kbVersion,
+            Key.keyboardRTL: isRTL,
+            Key.font: jsFont(fromFontDictionary: kbDict[Key.font] as? [AnyHashable: Any])
+          ]
+          if let oskFont = oskFont {
+            dict[Key.oskFont] = jsFont(fromFontDictionary: oskFont)
+          }
+          keyboardsDictionary[dictKey] = dict
+        }
+      }
+    }
+    updateUserKeyboardsList()
+  }
+
+  private func updateUserKeyboardsList() {
+    if keyboardsDictionary.isEmpty {
+      return
+    }
+    let userData = activeUserDefaults()
+
+    let lastVersion = userData.string(forKey: Key.engineVersion) ?? "1.0"
+    if compareVersions(lastVersion, sdkVersion) == .orderedSame {
+      return
+    }
+    userData.set(sdkVersion, forKey: Key.engineVersion)
+
+    guard var userKbList = userData.array(forKey: Key.userKeyboardsList) as? [[String: String]] else {
+      kmLog("No user keyboards to update", checkDebugPrinting: true)
+      return
+    }
+
+    for i in userKbList.indices {
+      let kbID = userKbList[i][Key.keyboardId]!
+      let langID = userKbList[i][Key.languageId]!
+      if var kb = keyboardsDictionary["\(langID)_\(kbID)"] {
+        kb[Key.keyboardVersion] = latestKeyboardFileVersion(withID: kbID)
+        kb[Key.customKeyboard] = "N"
+        userKbList[i] = kb
+      } else {
+        var kb = userKbList[i]
+        kb[Key.customKeyboard] = "Y"
+        userKbList[i] = kb
+      }
+    }
+    userData.set(userKbList, forKey: Key.userKeyboardsList)
+    userData.synchronize()
+  }
+
+  func keyboardIdForCurrentRequest() -> String? {
+    if let currentRequest = currentRequest {
+      let tmpStr = currentRequest.url.lastPathComponent
+      if tmpStr.hasSuffix(".js") {
+        return String(tmpStr.dropLast(3))
+      }
+      // TODO: Maybe don't return nil and just continue
+      return nil
+    } else if let downloadQueue = downloadQueue {
+      let kbInfo = downloadQueue.userInfo[Key.keyboardInfo]
+      if let dict = kbInfo as? [String: String] {
+        return dict[Key.keyboardId]
+      } else if let array = kbInfo as? [[String: String]] {
+        return array[0][Key.keyboardId]
+      }
+    }
+    return nil
+  }
+
+  func downloadKeyboard(forLanguageIndex languageIndex: Int, keyboardIndex: Int, isUpdate: Bool) {
+    guard let keyboard = keyboardInfo(forLanguageIndex: languageIndex, keyboardIndex: keyboardIndex) else {
```

Comment:
Why the refactor?  The original method was pretty small, and it appears the new helper function is only utilized here.

Reply:
Removing that helper function and just using `keyboard(for:)` instead.